### PR TITLE
Modify Hdfs monad to use tailcalls

### LINKED ISSUE #27: Use tail calls in resultant monad recursion to avoid stack overflows
As suggested in https://github.com/CommBank/omnitool/issues/27 I have a stack trace with about 1000 lines of:

````
at au.com.cba.omnia.permafrost.hdfs.Hdfs$$anon$1$$anonfun$rBind$1.apply(Hdfs.scala:338)
````

suggesting this is necessary for the Permafrost Hdfs resultant monad.

### PR DESCRIPTION
**Pull request by thsutton**
Closes #27

### CODE REVIEW

File: src/main/scala/au/com/cba/omnia/permafrost/hdfs/Hdfs.scala [hunk 1]
```diff
@@ -24,7 +26,12 @@ import au.com.cba.omnia.omnitool.{Result, ResultantMonad, ResultantOps, ToResult
  * HDFS operations use a hadoop configuration as context,
  * and produce a (potentially failing) result.
  */
-case class Hdfs[A](run: Configuration => Result[A])
+case class Hdfs[A](
+  val runTailRec: Configuration => TailRec[Result[A]]
```
Comment (rowandavies) [inline src/main/scala/au/com/cba/omnia/permafrost/hdfs/Hdfs.scala]: I guess you don’t need the `val` here.

File: src/main/scala/au/com/cba/omnia/permafrost/hdfs/Hdfs.scala [hunk 2]
```diff
@@ -33,11 +40,17 @@ case class Hdfs[A](run: Configuration => Result[A])
   */
 object Hdfs extends ResultantOps[Hdfs] with ToResultantMonadOps with HdfsROps[Hdfs] {
```
Comment (rowandavies) [inline src/main/scala/au/com/cba/omnia/permafrost/hdfs/Hdfs.scala]: Adding an `apply[A](run: Configuration => Result[A]): Hdfs[A]` would make this backward compatible, I think, aside from the case where there’s a pattern match against `Hdfs(…)` which seems unlikely.

This should mean that existing code using the original constructor will work, and you get the stack cleared each time you sequence `Hdfs` operations via `bind` or `rBind`.

**Comment by rowandavies** [comment]
Have you tested whether this resolves https://github.com/CommBank/AnalyticalWarehouse/issues/58 ?

**Comment by rowandavies** [comment]
Also, before merging this we should check that the tests pass for `maestro` and related dependencies.

File: src/main/scala/au/com/cba/omnia/permafrost/hdfs/Hdfs.scala (hunk 2 above)
Comment (thsutton) [inline src/main/scala/au/com/cba/omnia/permafrost/hdfs/Hdfs.scala]: This conflicts, after erasure, with the automatic `apply[A](run: Configuration => TailRec[Result[A]]): Hdfs[A]`. :-(

**Comment by thsutton** [comment]
@rowandavies No, I haven't tested whether it fixes the issue in Hydro but I'm running the Maestro tests now.

**Comment by thsutton** [comment]
This doesn't work anyway.

**Pull request closed without merging**